src/tests/suite/daos_dist_tx.c

```diff
@@ -0,0 +1,986 @@
+
+	ioreq_fini(&req);
+	MUST(daos_tx_close(th, NULL));
+}
+
+static void
+dtx_13(void **state)
+{
+	test_arg_t	*arg = *state;
+	const char	*dkey = dts_dtx_dkey;
+	const char	*akey = dts_dtx_akey;
+	char		 write_buf[DTX_IO_SMALL];
+	char		 fetch_buf[DTX_IO_SMALL];
+	daos_handle_t	 th = { 0 };
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+	int		 rc;
+
+	print_message("DTX status machnie\n");
+
+	arg->async = 0;
+	oid = dts_oid_gen(OC_RP_XSF, 0, arg->myrank);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_ARRAY, arg);
+	dts_buf_render(write_buf, DTX_IO_SMALL);
+
+	print_message("Open the TX1...\n");
+	MUST(daos_tx_open(arg->coh, &th, 0, NULL));
+
+	print_message("Commit the empty TX1...\n");
+	MUST(daos_tx_commit(th, NULL));
+
+	print_message("Commit the committed TX1...\n");
+	MUST(daos_tx_commit(th, NULL));
+
+	print_message("Abort the committed TX1...\n");
+	rc = daos_tx_abort(th, NULL);
+	assert_int_equal(rc, -DER_NO_PERM);
+
+	print_message("Restart the committed TX1...\n");
+	rc = daos_tx_restart(th, NULL);
+	assert_int_equal(rc, -DER_NO_PERM);
+
+	print_message("Update against the committed TX1...\n");
+	req.arg->expect_result = -DER_NO_PERM;
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	print_message("Fetch against the committed TX1...\n");
+	req.arg->expect_result = 0;
+	lookup_single(dkey, akey, 0, fetch_buf, DTX_IO_SMALL, th, &req);
+	assert_int_equal(req.iod[0].iod_size, 0);
+
+	print_message("Close the TX1...\n");
+	MUST(daos_tx_close(th, NULL));
+
+	print_message("Open the TX2...\n");
+	MUST(daos_tx_open(arg->coh, &th, 0, NULL));
+
+	print_message("Update via the TX2...\n");
+	req.arg->expect_result = 0;
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	print_message("Restart the TX2...\n");
+	MUST(daos_tx_restart(th, NULL));
+
+	print_message("Restart the TX2 again...\n");
+	MUST(daos_tx_restart(th, NULL));
+
+	print_message("Abort the TX2...\n");
+	MUST(daos_tx_abort(th, NULL));
+
+	print_message("Abort the TX2 again...\n");
+	MUST(daos_tx_abort(th, NULL));
+
+	print_message("Commit the aborted TX2...\n");
+	rc = daos_tx_commit(th, NULL);
+	assert_int_equal(rc, -DER_NO_PERM);
+
+	print_message("Update against the aborted TX2...\n");
+	req.arg->expect_result = -DER_NO_PERM;
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	print_message("Fetch against the aborted TX2...\n");
+	req.arg->expect_result = -DER_NO_PERM;
+	lookup_single(dkey, akey, 0, fetch_buf, DTX_IO_SMALL, th, &req);
+
+	print_message("Close the TX2...\n");
+	MUST(daos_tx_close(th, NULL));
+
+	ioreq_fini(&req);
+}
+
+static void
+dtx_14(void **state)
+{
+	test_arg_t	*arg = *state;
+	const char	*dkey = dts_dtx_dkey;
+	const char	*akey = dts_dtx_akey;
+	const char	*akey2 = "tmp_akey";
+	char		 write_buf[DTX_IO_SMALL];
+	char		 fetch_buf[DTX_IO_SMALL];
+	daos_handle_t	 th = { 0 };
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+	int		 rc;
+
+	print_message("DTX restart because of conflict with others\n");
```

Comment:
I think it would be good to create a simple workload between many ranks that actually generates conflicts rather than just passing param setting to simulate artificial conflicts. after all we do need to have conflict detection properly tested. 

But maybe this is not the purpose of this patch / test and you are only testing the dtx component and not the mvcc one?

Reply:
Right, these test cases are only for object level DTX conflict handle. As for how to detect conflict, it is MVCC/VOS level logic, will be tested by related VOS test cases.